### PR TITLE
Add solver dependent callbacks

### LINKED ISSUE #58: Solver dependent callbacks
Add to MOI wrapper

### PR DESCRIPTION
**Pull request by rafabench**
Close #58 
Added MOI heuristics and user cuts callbacks
Also added a callback to print log in Windows

### CODE REVIEW

**Comment by frapac** [comment]
Thanks for adding this to Xpress.jl! 
In my experience, we have to be careful when using lazy callbacks with Xpress. Most of the time I have to deactivate the presolve to avoid annoying error. 
E.g., testing the new callbacks you just added, I get 
``` 
?405 Error: Invalid column number passed to XPRSstorecuts.
Column number 0 is invalid
ERROR: LoadError: XpressError(32): Subroutine not completed successfully, possibly due to invalid argument. Unable to call `Xpress.storecuts`:

405 Error: Invalid column number passed to XPRSstorecuts.
Column number 0 is invalid
```
as the presolve removes all the columns on this simple model. The problem here is that we do not know on the MOI side which variables is using Xpress after the presolve.

**Comment by joaquimg** [comment]
thanks @frapac !
This error is raised in some test already here?
Or is it something you built?

**Comment by frapac** [comment]
In this PR, the tests are carefully written and set explicitly [the right set of options](https://github.com/JuliaOpt/Xpress.jl/blob/rb/features/test/MathOptInterface/MOI_callbacks.jl#L14-L18). I am wondering what is the correct way to signal that to the end user

* by setting explicitly `PRESOLVE` to `0` in the MOI wrapper
* or by adding more documentation about the callbacks in the README (a warning issued in the MOI wrapper could also do the job)

This error was raised by something I built (basically a sparse logistic regression with lazy callbacks implemented on top of Xpress). Unfortunately, I still have some trouble to make them work as expected... If you are interested, you could find the source [here](https://github.com/frapac/LogisticOptTools.jl/blob/master/examples/sparse_regression.jl)

File: test/MathOptInterface/MOI_callbacks.jl [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+                floor.(x_vals)
+            ) == MOI.HEURISTIC_SOLUTION_UNKNOWN
+            global callback_called = true
+        end)
+        @test MOI.supports(model, MOI.HeuristicCallback())
+        MOI.optimize!(model)
+        @test callback_called
+    end
+    #=
+    @testset "LazyConstraint" begin
```
Comment (joaquimg) [inline test/MathOptInterface/MOI_callbacks.jl:252]: we need to add these submission errors

**Review by joaquimg — COMMENTED** [review]
som tweaks

File: src/helper.jl [hunk 2]
```diff
@@ -53,21 +53,30 @@ Base.unsafe_convert(T::Type{Ptr{Nothing}}, t::CWrapper) = (t.ptr == C_NULL) ? th
...
-        ref = Ref{Lib.XPRSprob}()
-        createprob(ref)
-        @assert ref[] != C_NULL "Failed to create XpressProblem. Received null pointer from Xpress C interface."
-        p = new(ref[], Any[], 0.0, logfile)
-        if logfile != ""
-            setlogfile(p, logfile)
+    function XpressProblem(ptr::Lib.XPRSprob; finalize_env::Bool = true, logfile = "")
+        model = new(ptr, Any[], finalize_env, 0.0, logfile)
+        if finalize_env
+            finalize(() -> destroyprob(model))
```
Comment (kdheepak) [inline src/helper.jl:62]: See https://github.com/jump-dev/Xpress.jl/pull/100/commits/93d3ecec271315aae60d7e3cfa07e6effd62c450 for correct use of `finalizer`, the old code was incorrect.

**Pull request merged**